### PR TITLE
Add Dynamic Repeat 

### PR DESCRIPTION
**Pull request by calebmkim**
Still a draft rn. 

(Edit, ready for review): 
Adds a static repeat operator 

Works by transforming 
```
repeat num_repeats {repeat_body}
```
into 
```
initialize cond_reg = 1; idx = 0; 
while cond_reg.out {
  body; 
  cond_reg = (idx + 1 < num_repeats) 
  idx +=1; 
}
```

I also made some changes to the domination analysis: I realized that there was 
an edge case (`repeats` w/ `num_repeats=0` should be ignored in the dominator map) and added 
a test case. 

Also a correctness test case for compiling static repeats. (I turned `static-promotion` pass
off for this test case just to be able to test the compilation of dynamic repeats)

### CODE REVIEW

**Comment by rachitnigam** [comment]
Awesome! Does `static-promotion` also support promotion using it now or is that another PR?

**Comment by calebmkim** [comment]
Yep, it promotes to `static repeat` if the body is static.

**Comment by calebmkim** [comment]
(Obviously not actionable in this PR). Next step could be to get CIRCT Calyx to support `repeat` statements, and then making the MLIR->Calyx flows supporting `repeat` statements.

File: calyx-frontend/src/ast.rs [hunk 1]
```diff
@@ -436,6 +447,7 @@ impl Control {
             Control::Par { attributes, .. } => attributes,
             Control::If { attributes, .. } => attributes,
             Control::While { attributes, .. } => attributes,
+            Control::DynamicRepeat { attributes, .. } => attributes,
```
Comment (rachitnigam) [inline calyx-frontend/src/ast.rs:450]: Why is this called dynamic repeat instead of just `Repeat`?

File: calyx-frontend/src/parser.rs [hunk 2]
```diff
@@ -1001,6 +1001,18 @@ impl CalyxParser {
         ))
     }
 
+    fn dynamic_repeat_stmt(input: Node) -> ParseResult<ast::Control> {
```
Comment (rachitnigam) [inline calyx-frontend/src/parser.rs:1004]: Both the repeat statements return an `ast::Control` and so should have one rule to handle them. Seems like the only difference in the syntax is the `static_word` stuff

File: calyx-opt/src/analysis/live_range_analysis.rs [hunk 3]
```diff
@@ -1373,6 +1391,21 @@ impl LiveRangeAnalysis {
...
+                    self.build_live_ranges(body, alive, gens, kills);
+                if *num_repeats == 1 {
+                    // only one execution, so we don't need to iterate through
+                    // twice
+                    (a, g, k)
+                } else {
+                    // need to feed the live nodes on the output of the body
+                    // back into the body to get correct live range analysis
+                    self.build_live_ranges(body, a, g, k)
+                }
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/live_range_analysis.rs:1407]: I'm a little iffy about the special handling of the `repeat 0` and `repeat 1` cases in the analyses. To me, `collapse-control` should be the thing that gets rid of trivial repeat statements and the logic in the analysis should be generic. This does have the problem that users can get potentially unpredictable results if they forget to call that pass before running the analysis. 

Maybe @sampsyo can chime on this as well?

File: calyx-opt/src/analysis/reaching_defns.rs [hunk 4]
```diff
@@ -636,6 +637,50 @@ fn build_reaching_def(
             en.group.borrow().name(),
         ),
         ir::Control::Empty(_) => (reach, killed),
+        ir::Control::Repeat(ir::Repeat { body, .. }) => {
+            // (XXX) If we start using this analysis, we should think more carefully
+            // about how exactly to handle repeats. This code basically just copies
+            // what while loops do (except for its dealing with cond ports).
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/reaching_defns.rs:643]: We already use this analysis so I'm not sure what the comment means. Also, let's abstract this code a bit so that we're not literally duplicating the code. Finally, for `TODO`, `XXX`, or `NOTE` comments, put your name in there so we know who wrote it: `XXX(rachit): ...`

File: interp/src/flatten/flat_ir/control/translator.rs [hunk 5]
```diff
@@ -612,6 +612,9 @@ impl FlattenTree for cir::Control {
             cir::Control::Static(_) => {
                 todo!("The interpreter does not support static control yet")
             }
+            cir::Control::Repeat(_) => {
+                todo!("The interpreter does not support repeat yet")
```
Comment (rachitnigam) [inline interp/src/flatten/flat_ir/control/translator.rs:616]: @EclecticGriffin should we make an issue for this? Seems like it would be a "good first issue" since the logic for loops is already there

**Review by rachitnigam — COMMENTED** [review]
Mostly looks good! Just one top-level comment about how we should handle the special cases of `repeat 0` and `repeat 1`. It seems a bit weird that the analysis have to deal them. I was expecting that `collapse-control` could be responsible for simplifying them instead and having the analyses use generic code. Wanted some feedback on this from @sampsyo as well.

File: interp/src/flatten/flat_ir/control/translator.rs (hunk 5 above)
Comment (EclecticGriffin) [reply interp/src/flatten/flat_ir/control/translator.rs:616]: Yeah if you wouldn't mind opening an issue for it! I don't think it would be hard to implement in the existing infra. The flat stuff is it's own other world not worth considering yet

File: calyx-opt/src/passes/static_promotion.rs [hunk 6]
```diff
@@ -775,4 +775,34 @@ impl Visitor for StaticPromotion {
...
+        &mut self,
+        s: &mut ir::Repeat,
+        _comp: &mut ir::Component,
+        _sigs: &LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        if s.body.is_static() {
+            // need to do this weird thing to get the repeat body
+            let empty = Box::new(ir::Control::empty());
+            let repeat_body = std::mem::replace(&mut s.body, empty);
```
[the remark this answers is not in the document]
Comment (EclecticGriffin) [reply calyx-opt/src/passes/static_promotion.rs:790]: I would recommend this!

**Comment by calebmkim** [comment]
Yeah, I see what you're saying, and was actually thinking along the same lines as you. Maybe I could do the following changes: 

- Implement the `repeat` simplification in the canonicalization pass, and get rid of the special cases for the analyses. 
- For the domination stuff, maybe we should throw an error if we see the "repeat 0" case? Because the analysis is literally incorrect if we treat "repeat 0" like the other "repeat" statements.

File: calyx-ir/src/control.rs [hunk 7]
```diff
@@ -319,19 +342,6 @@ pub enum StaticControl {
...
-    }
-
-    pub fn get_latency(&self) -> Option<StaticLatency> {
-        match self {
-            Control::Static(sc) => Some(sc.get_latency()),
-            _ => None,
-        }
-    }
-}
-
```
Comment (calebmkim) [inline calyx-ir/src/control.rs:334]: For some reason there was `impl Control` twice in the same file. I decided to change this bc it was confusing.

File: calyx-opt/src/analysis/live_range_analysis.rs (hunk 3 above)
Comment (sampsyo) [reply calyx-opt/src/analysis/live_range_analysis.rs]: Yeah, that's a good point… it's not a huge sin, if handling the n=1 case in concert with the general case is just too annoying to implement! But as a somewhat weak first-order principle, it keeps things a little simpler if we try to adhere to "division of labor" when it's not in conflict with other goals.

**Review by sampsyo — COMMENTED** [review]
Heck yes; this looks great!! One tiny suggestion: maybe add a quick little section to the language reference to describe `repeat` loops?
https://docs.calyxir.org/lang/ref.html

File: calyx-opt/src/analysis/live_range_analysis.rs (hunk 3 above)
Comment (calebmkim) [reply calyx-opt/src/analysis/live_range_analysis.rs]: Ok, what I've done for now is added dynamic `repeat 0/1` as a case for `collapse-control`. 
I also changed the default pipeline to run `collapse-control` twice: once at the beginning of the `pre-opt` and once at the end. 
Live Range Analysis now doesn't make a special case for n=1 (it relies on `collapse-control` to handle these speical cases). 
For `DominationAnalysis`, I still handle the n=0 case. The reason why I want to keep it is that the domination analysis would actually be incorrect if we didn't especially handle the n=0 case... things in a repeat n=0 statement shouldn't dominate anything, and if we say they do, then that's wrong. 

Either way, lmk if you think I should change any of these. I get that what I'm doing rn may not nec be the cleanest solution.

**Comment by rachitnigam** [comment]
This is looking good to me! Let's merge it when you're ready @calebmkim!
